File: go/0853_car_fleet.go
package main

import "sort"

func carFleet(target int, position []int, speed []int) int {
	carlen := len(position)
	if carlen <= 1 {
		return carlen
	}

	cars := make([]*Car, carlen)
	for i := 0; i < carlen; i++ {
		cars[i] = &Car{Position: position[i], Speed: speed[i]}
	}

	sort.Slice(cars, func(i, j int) bool {
		if cars[i].Position >= cars[j].Position {
			return true
		}
		return false
	})

	fleet := 1
	carInFront := cars[0]
	for i := 1; i < carlen; i++ {
		if float32(target-carInFront.Position)/float32(carInFront.Speed) <
			float32(target-cars[i].Position)/float32(cars[i].Speed) {
			fleet++
			carInFront = cars[i]
		}
	}
	return fleet
}

type Car struct {
	Position int
	Speed    int
}

File: python/0853_car_fleet.py
class Solution:
    def carFleet(self, target: int, position: List[int], speed: List[int]) -> int:
        cars = []
        for p, s in zip(position, speed):
            cars.append(Car(p, s))
        
        if not cars:
            return 0
        
        cars.sort()
        
        fleet = len(cars)
        prevCar = cars[-1]
        prevCarDur = (target - prevCar.pos) / prevCar.speed
        for i in range(len(cars) - 2, -1, -1):
            curCar = cars[i]
            curCarDur = (target - curCar.pos) / curCar.speed

            if curCarDur <= prevCarDur:

                fleet -= 1
                continue
            prevCar = curCar
            prevCarDur = curCarDur

        return fleet
        

class Car:
    def __init__(self, pos, speed):
        self.pos = pos
        self.speed = speed
        
    def __lt__(self, other):
        if self.pos < other.pos:
            return True
        elif self.pos > other.pos:
            return False
        return self.speed > other.speed